subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 #if defined(CONFIG_DNS_RESOLVER)
 		case DHCPV4_OPTIONS_DNS_SERVER: {
 			struct dns_resolve_context *ctx;
-			struct sockaddr_in dns;
-			const struct sockaddr *dns_servers[] = {
-				(struct sockaddr *)&dns, NULL
-			};
+			struct sockaddr_in dnses[CONFIG_DNS_RESOLVER_MAX_SERVERS];
+			const struct sockaddr *dns_servers[CONFIG_DNS_RESOLVER_MAX_SERVERS];
+
+			for (uint8_t i = 0; i < CONFIG_DNS_RESOLVER_MAX_SERVERS; i++) {
+				dns_servers[i] = (struct sockaddr *)&dnses[i];
+			}
 			int status;
 
 			/* DNS server option may present 1 or more
 			 * addresses. Each 4 bytes in length. DNS
 			 * servers should be listed in order
-			 * of preference.  Hence we choose the first
-			 * and skip the rest.
+			 * of preference. Hence how many we parse
+			 * depends on CONFIG_DNS_RESOLVER_MAX_SERVERS
 			 */
-			if (length % 4 != 0U) {
+			const uint8_t addr_size = 4U;
+
+			if (length % addr_size != 0U) {
 				NET_ERR("options_dns, bad length");
 				return false;
 			}
 
-			(void)memset(&dns, 0, sizeof(dns));
+			(void)memset(&dnses, 0, sizeof(dnses));
 
-			if (net_pkt_read(pkt, dns.sin_addr.s4_addr, 4) ||
-			    net_pkt_skip(pkt, length - 4U)) {
+			const uint8_t provided_servers_cnt = length/addr_size;
+			uint8_t dns_servers_cnt = 0;
+
+			if (provided_servers_cnt > CONFIG_DNS_RESOLVER_MAX_SERVERS) {
+				NET_WARN("DHCP server provided more DNS servers than can be saved");
+				dns_servers_cnt = CONFIG_DNS_RESOLVER_MAX_SERVERS;
+			} else {
+				for (uint8_t i = provided_servers_cnt;
+					i < CONFIG_DNS_RESOLVER_MAX_SERVERS; i++) {
```

Comment:
Please align with the opening bracket, so that the if condition does not blend with the code block.